File: instrumentation/android/appmap/src/main/kotlin/com/example/appmap/AppMapDeepLink.kt
// AppMapKit (Android) — test-only deep links (01 R5).
//
// Scheme: <scheme>://<screen_id>[?fixture=<name>&k=v…], `appmap` by default and configurable per
// app via AppMapDeepLink.scheme (issue #25: two instrumented apps on one device otherwise collide).
// Parsing is pure Kotlin so it unit-tests on the JVM. Everything is gated on
// BuildConfig.APP_MAP_DEBUG; the intent filter itself exists only in src/debug/AndroidManifest.xml.
package com.example.appmap

import android.net.Uri
import android.util.Log
import java.net.URLDecoder

/** A parsed app-map deep link. [params] never contains `fixture`. */
data class AppMapRoute(
    val screenId: String,
    val fixture: String? = null,
    val params: Map<String, String> = emptyMap(),
)

object AppMapDeepLink {
    /**
     * The default scheme. Every app-map app used to register exactly this, which meant two
     * instrumented apps on one device both claimed it and the OS delivered `appmap://<screen>` —
     * and the `?fixture=` it carries — to whichever it liked, while the app-scoped capture kept
     * describing the other one and timed out (issue #25).
     *
     * Set [scheme] from Application.onCreate in debug builds when more than one app-map app can be
     * installed at once, and declare the SAME value as `deep_link_scheme` in
     * `app-map/android/manifest.yaml` and as the `android:scheme` of the debug intent-filter.
     * `appmap-<last component of the application id>` is the conventional choice.
     */
    const val DEFAULT_SCHEME = "appmap"

    @Volatile
    var scheme: String = DEFAULT_SCHEME
        set(value) {
            val trimmed = value.trim().lowercase()
            require(SCHEME_PATTERN.matches(trimmed)) { "app-map deep link scheme must match ${SCHEME_PATTERN.pattern}, got '$value'" }
            field = trimmed
        }

    /** Kept for source compatibility with integrations written before the scheme was configurable. */
    @Deprecated("read `scheme`; it is configurable per app (issue #25)", ReplaceWith("scheme"))
    const val SCHEME = DEFAULT_SCHEME

    /** RFC 3986 `scheme`, narrowed the same way the manifest schema narrows it. */
    private val SCHEME_PATTERN = Regex("^[a-z][a-z0-9+.-]{0,63}$")
    private const val TAG = "app-map"

    /** Screen ids are bare snake_case names (01 R2; app-map/schema/ids.schema.json); the marker is `screen.<id>`. */
    private val SCREEN_ID = Regex("^[a-z][a-z0-9_]*$")

    /** The app's real router (01 R5). Installed from Application.onCreate in debug builds. */
    fun interface Router {
        fun route(route: AppMapRoute)
    }

    @Volatile
    private var router: Router? = null

    fun installRouter(router: Router) {
        if (!BuildConfig.APP_MAP_DEBUG) return
        this.router = router
    }

    /** Returns null in Release builds and for anything that is not a well-formed app-map link. */
    fun parse(uri: String): AppMapRoute? {
        if (!BuildConfig.APP_MAP_DEBUG) return null
        return parseUnchecked(uri)
    }

    fun parse(uri: Uri): AppMapRoute? = parse(uri.toString())

    /**
     * Applies the fixture (if any) via [AppMapFixtureRegistry], then hands the route to the
     * installed router. Returns false when disabled or no router is installed; fixture failures
     * propagate as [AppMapFixtureException].
     */
    fun dispatch(route: AppMapRoute): Boolean {
        if (!BuildConfig.APP_MAP_DEBUG) return false
        val target = router
        if (target == null) {
            Log.e(TAG, "$scheme://${route.screenId} received but no router installed (AppMapDeepLink.installRouter)")
            return false
        }
        route.fixture?.let { AppMapFixtureRegistry.apply(it) }
        target.route(route)
        return true
    }

    // Not gated: exercised directly by unit tests of the grammar. Hand-rolled because
    // java.net.URI returns a null host for names containing '_'.
    internal fun parseUnchecked(raw: String): AppMapRoute? {
        val text = raw.trim()
        val prefix = "$scheme://"
        if (!text.regionMatches(0, prefix, 0, prefix.length, ignoreCase = true)) return null
        val rest = text.substring(prefix.length).substringBefore('#')

        val queryIndex = rest.indexOf('?')
        val hostPart = (if (queryIndex >= 0) rest.substring(0, queryIndex) else rest).trimEnd('/')
        val query = if (queryIndex >= 0) rest.substring(queryIndex + 1) else ""
        if (!SCREEN_ID.matches(hostPart)) return null

        val params = LinkedHashMap<String, String>()
        for (pair in query.split('&')) {
            if (pair.isEmpty()) continue
            val eq = pair.indexOf('=')
            val key = decode(if (eq >= 0) pair.substring(0, eq) else pair)
            val value = decode(if (eq >= 0) pair.substring(eq + 1) else "")
            if (key.isNotEmpty()) params[key] = value
        }
        val fixture = params.remove("fixture")
        return AppMapRoute(hostPart, fixture, params)
    }

    private fun decode(s: String): String = URLDecoder.decode(s, "UTF-8")
}

File: instrumentation/ios/AppMapKit/Sources/AppMapKit/AppMapDeepLink.swift
// AppMapKit — test-only deep links (01 R5).
//
// Scheme: <scheme>://<screen_id>[?fixture=<name>&k=v…], `appmap` by default and configurable per
// app via AppMapDeepLink.scheme (issue #25: two instrumented apps on one device otherwise collide).
// The parser produces an `AppMapRoute`; the app's *real* router consumes it so the resulting
// state is genuine, not a bare view. Nothing in this file exists outside APP_MAP_DEBUG: the
// `#if` leaves no symbol behind, which is what 01 §4 asks a Release unit test to verify
// (see Tests/AppMapKitTests/AppMapReleaseGateTests.swift).

import Foundation

#if APP_MAP_DEBUG
/// A parsed app-map deep link. `params` never contains `fixture`.
public struct AppMapRoute: Equatable, Sendable {
    public let screenID: String
    public let fixture: String?
    public let params: [String: String]

    public init(screenID: String, fixture: String? = nil, params: [String: String] = [:]) {
        self.screenID = screenID
        self.fixture = fixture
        self.params = params
    }
}

public enum AppMapDeepLink {
    /// The default scheme.
    ///
    /// Every app-map app used to register exactly this, which meant two instrumented apps on one
    /// device both claimed it and iOS delivered `appmap://<screen>` — and the `?fixture=` it
    /// carries — to whichever it liked, while the app-scoped capture kept describing the other one
    /// and timed out (issue #25).
    public static let defaultScheme = "appmap"

    /// The scheme THIS app registers. Set it once at launch in debug builds when more than one
    /// app-map app can be installed at once, and declare the SAME value as `deep_link_scheme` in
    /// `app-map/ios/manifest.yaml` and in the Debug target's `CFBundleURLTypes`.
    /// `appmap-<last component of the bundle id>` is the conventional choice.
    ///
    ///     AppMapDeepLink.scheme = "appmap-pokedexteams"
    ///
    /// Setting an invalid scheme is a programmer error and traps: an app that answers a scheme the
    /// map does not name is the failure this exists to prevent, and it is silent at runtime.
    public static var scheme: String {
        get { storedScheme }
        set {
            let normalized = newValue.trimmingCharacters(in: .whitespaces).lowercased()
            precondition(
                normalized.range(of: schemePattern, options: .regularExpression) != nil,
                "app-map deep link scheme must be a lowercase URL scheme, got '\(newValue)'"
            )
            storedScheme = normalized
        }
    }

    private static var storedScheme = defaultScheme
    /// RFC 3986 `scheme`, narrowed the same way `manifest.schema.json` narrows `deep_link_scheme`.
    private static let schemePattern = "^[a-z][a-z0-9+.-]{0,63}$"
    /// Screen ids are bare snake_case names (01 R2; app-map/schema/ids.schema.json); the marker is `screen.<id>`.
    private static let screenIDPattern = "^[a-z][a-z0-9_]*$"

    public static func parse(_ url: URL) -> AppMapRoute? {
        guard let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else { return nil }
        return parse(components)
    }

    public static func parse(_ string: String) -> AppMapRoute? {
        guard let components = URLComponents(string: string) else { return nil }
        return parse(components)
    }

    private static func parse(_ components: URLComponents) -> AppMapRoute? {
        guard components.scheme?.lowercased() == storedScheme else { return nil }
        guard let host = components.host, !host.isEmpty else { return nil }
        guard components.path.isEmpty || components.path == "/" else { return nil }
        guard host.range(of: screenIDPattern, options: .regularExpression) != nil else { return nil }

        var params: [String: String] = [:]
        for item in components.queryItems ?? [] where !item.name.isEmpty {
            params[item.name] = item.value ?? ""
        }
        let fixture = params.removeValue(forKey: "fixture")
        return AppMapRoute(screenID: host, fixture: fixture, params: params)
    }
}

/// Applies the optional fixture, then hands the route to the app's real router.
///
/// Exposed to the Objective-C runtime under a fixed name so a Release build can be probed for
/// its absence without a compile-time reference: `NSClassFromString("AppMapDeepLinkHandler")`.
///
///     // Debug-only wiring in the App / SceneDelegate:
///     let handler = AppMapDeepLinkHandler(fixtures: DebugFixtures()) { route in
///         AppRouter.shared.navigate(to: route.screenID, params: route.params)
///     }
///     .onOpenURL { url in _ = handler.handle(url) }
@objc(AppMapDeepLinkHandler)
public final class AppMapDeepLinkHandler: NSObject {
    public typealias Router = @MainActor (AppMapRoute) -> Void

    private let fixtures: AppMapFixtures?
    private let route: Router

    public init(fixtures: AppMapFixtures? = nil, route: @escaping Router) {
        self.fixtures = fixtures
        self.route = route
        super.init()
    }

    /// Returns `false` when `url` is not an app-map link so the caller can fall through to its
    /// production URL handling. Fixture application and routing run asynchronously on the main
    /// actor; a failing fixture logs and does not route.
    @discardableResult
    public func handle(_ url: URL) -> Bool {
        guard let parsed = AppMapDeepLink.parse(url) else { return false }
        let fixtures = self.fixtures
        let route = self.route
        Task { @MainActor in
            if let name = parsed.fixture {
                guard let fixtures else {
                    AppMapLog.error("fixture '\(name)' requested but no AppMapFixtures installed")
                    return
                }
                do {
                    try await fixtures.apply(name: name)
                } catch {
                    AppMapLog.error("fixture '\(name)' failed: \(error)")
                    return
                }
            }
            route(parsed)
        }
        return true
    }
}
#endif
